libafl/src/corpus/mod.rs

```diff
@@ -36,26 +69,96 @@ pub trait Corpus: UsesInput + serde::Serialize + for<'de> serde::Deserialize<'de
     }
 
     /// Add an entry to the corpus and return its index
-    fn add(&mut self, testcase: Testcase<Self::Input>) -> Result<usize, Error>;
+    fn add(&mut self, testcase: Testcase<Self::Input>) -> Result<CorpusId, Error>;
 
     /// Replaces the testcase at the given idx, returning the existing.
     fn replace(
         &mut self,
-        idx: usize,
+        idx: CorpusId,
         testcase: Testcase<Self::Input>,
     ) -> Result<Testcase<Self::Input>, Error>;
 
     /// Removes an entry from the corpus, returning it if it was present.
-    fn remove(&mut self, idx: usize) -> Result<Option<Testcase<Self::Input>>, Error>;
+    fn remove(&mut self, idx: CorpusId) -> Result<Testcase<Self::Input>, Error>;
 
     /// Get by id
-    fn get(&self, idx: usize) -> Result<&RefCell<Testcase<Self::Input>>, Error>;
+    fn get(&self, idx: CorpusId) -> Result<&RefCell<Testcase<Self::Input>>, Error>;
 
     /// Current testcase scheduled
-    fn current(&self) -> &Option<usize>;
+    fn current(&self) -> &Option<CorpusId>;
 
     /// Current testcase scheduled (mutable)
-    fn current_mut(&mut self) -> &mut Option<usize>;
+    fn current_mut(&mut self) -> &mut Option<CorpusId>;
+
+    /// Get the next corpus id
+    fn next(&self, idx: CorpusId) -> Option<CorpusId>;
+
+    /// Get the prev corpus id
+    fn prev(&self, idx: CorpusId) -> Option<CorpusId>;
+
+    /// Get the first inserted corpus id
+    fn first(&self) -> Option<CorpusId>;
+
+    /// Get the last inserted corpus id
+    fn last(&self) -> Option<CorpusId>;
+
+    /// An iterator over very active corpus id
+    fn ids(&self) -> CorpusIdIterator<'_, Self> {
+        CorpusIdIterator {
+            corpus: self,
+            cur: self.first(),
+            cur_back: self.last(),
+        }
+    }
+
+    /// Get a random corpus id
+    fn random_id(&self, next_random: u64) -> CorpusId {
```

Comment:
The name here isn't really correct, it's not random at all (only the parameter is)
Maybe `id_from_u64` or something?

Reply:
I'm undecided about the function not just the name. Ideally, we should pass a &mut Rand, but as the prng is in the State and the corpus too it would be a mutable ref to get the prng and a normal ref to get the corpus with the borrow checker complaining about it.  see the random_corpus_id macro, it works but it's not the cleanest solution